File: docs/audit/2026-09-12/repro/asset-tools-36.py
"""Audit-only isolated checks. Product scripts are read, never called with product outputs."""
from __future__ import annotations
import ast
import contextlib
import hashlib
import io
import json
import math
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import wave
from unittest.mock import patch
import numpy as np
from PIL import Image

ROOT = Path(__file__).resolve().parents[4]
COCOS = ROOT / "work/guandan-cocos"
def digest(data):
    return hashlib.sha256(data).hexdigest()
def module(relative):
    path = COCOS / relative
    ns = {"__file__": str(path), "__name__": "audit_only_no_entrypoint"}
    exec(compile(path.read_text(), str(path), "exec"), ns)
    return ns

# Existing modules are guarded by __main__; no save/build entrypoints are invoked.
timer = module("scripts/extract-timer-frame.py")
bomb = module("scripts/generate-bomb-vfx-assets.py")
docpath = COCOS / "scripts/generate-gameplay-audio-effects-docx.py"
doctree = ast.parse(docpath.read_text())
compile(doctree, str(docpath), "exec")
assert any(isinstance(n, ast.If) and "__main__" in ast.unparse(n.test) for n in doctree.body)
pictures = sorted({n.value for n in ast.walk(doctree) if isinstance(n, ast.Constant)
                   and isinstance(n.value, str) and n.value.endswith(".png")})
assert len(pictures) == 8
assert all((COCOS / "docs/assets/game-flow" / name).is_file() for name in pictures)
# Deliberately do not build the historical Word document, so no product docx is replaced.
doc = {"syntax": True, "existingPictureInputs": len(pictures), "buildInvoked": False,
       "hardcodedHistoricalDate": "2026-08-04"}

# Exterior fill must not erase white details enclosed inside a colored object.
rgb = np.full((9, 9, 3), 220, dtype=np.uint8)
rgb[2:7, 2:7] = [120, 80, 10]
rgb[3:6, 3:6] = 255
outside = timer["exterior_background"](rgb)
assert np.count_nonzero(outside) == 56
assert not outside[4, 4]
solid = np.full((5, 5, 3), [200, 210, 200], dtype=np.uint8)
assert not timer["exterior_background"](solid).any()
assert timer["exterior_background"](np.full((5, 5, 3), 200, dtype=np.uint8)).all()
synthetic = np.zeros((9, 9, 4), dtype=np.uint8)
synthetic[2:7, 2:7] = [240, 0, 0, 255]
resized = timer["resize_premultiplied"](Image.fromarray(synthetic), (27, 27))
pixels = np.asarray(resized)
assert pixels.shape == (27, 27, 4)
assert (pixels[:, :, 3] == 0).any() and (pixels[:, :, 3] == 255).any()
assert (pixels[:, :, 3] > 0).any()
assert not pixels[:, :, 1:3].any()
# Eight-bit premultiplication amplifies rounding at nearly transparent pixels.
# Check color retention where alpha >= 64; fully transparent pixels need no color bound.
assert (pixels[:, :, 0][pixels[:, :, 3] >= 64] >= 230).all()

# Replace only the final Pillow save boundary; original extraction executes on synthetic input.
captured = []
save_original = Image.Image.save
def capture_save(image, target, *args, **kwargs):
    buffer = io.BytesIO()
    save_original(image, buffer, format="PNG", **kwargs)
    captured.append({"image": image.copy(), "data": buffer.getvalue(), "target": str(target)})
with tempfile.TemporaryDirectory(prefix="audit-assets36-") as td:
    scratch = Path(td)
    source = scratch / "synthetic-timer.png"
    Image.fromarray(rgb).save(source)
    target = scratch / "output" / "timer.png"
    with patch.object(Image.Image, "save", capture_save), contextlib.redirect_stdout(io.StringIO()):
        timer["extract"](source, target, 64)
    assert not target.exists() and captured[-1]["image"].size == (64, 64)
    blank = scratch / "blank.png"
    Image.new("RGB", (9, 9), (220, 220, 220)).save(blank)
    try:
        timer["extract"](blank, scratch / "blank-output.png", 64)
        raise AssertionError("blank source unexpectedly accepted")
    except RuntimeError as error:
        assert "No foreground" in str(error)
    assert not (scratch / "blank-output.png").exists()
timer_result = {"borderPixelsRemoved": 56, "enclosedWhitePreserved": True,
                "premultipliedNoGreenBlueBleed": True, "syntheticExtract": "64x64",
                "blankRejected": True, "productImagesWritten": 0}

# Bleeding may only change RGB, never alpha or pixels outside the specified distance.
bleed_input = Image.new("RGBA", (9, 9))
bleed_input.putpixel((4, 4), (240, 70, 10, 255))
bled = bomb["transparent_rgb_bleed"](bleed_input, distance=2)
assert bled.getchannel("A").tobytes() == bleed_input.getchannel("A").tobytes()
assert bled.getpixel((2, 2)) == (240, 70, 10, 0)
assert bled.getpixel((1, 1)) == (0, 0, 0, 0)
assert bleed_input.getpixel((2, 2)) == (0, 0, 0, 0)

class MemoryFile:
    def __init__(self, name): self.name = name; self.data = None
    def read_bytes(self): assert self.data is not None; return self.data
class MemoryDirectory:
    def __init__(self): self.files = {}
    def __truediv__(self, name):
        return self.files.setdefault(name, MemoryFile(name))
memory_output = MemoryDirectory()
def bomb_save(image, target, *args, **kwargs):
    assert isinstance(target, MemoryFile)
    output = io.BytesIO()
    save_original(image, output, *args, **kwargs)
    target.data = output.getvalue()

assets = []
for name, (size, generator, transparent) in bomb["ASSETS"].items():
    first, second = generator(), generator()
    assert first.mode == "RGBA" and first.size == size
    assert first.tobytes() == second.tobytes()
    with patch.object(Image.Image, "save", bomb_save), contextlib.redirect_stdout(io.StringIO()):
        sha = bomb["save_asset"](name, first, memory_output, transparent)
    rendered = memory_output.files[name].data
    existing = COCOS / "assets/game-assets/effects/bomb-v1" / name
    with Image.open(existing) as bundled:
        bundled_rgba = bundled.convert("RGBA")
        memory_rgba = Image.open(io.BytesIO(rendered)).convert("RGBA")
        pixel_match = bundled_rgba.size == memory_rgba.size and bundled_rgba.tobytes() == memory_rgba.tobytes()
    assets.append({"name": name, "size": size, "deterministicPixels": True,
                   "generatedSha256": sha, "existingSha256": digest(existing.read_bytes()),
                   "byteMatch": rendered == existing.read_bytes(), "pixelMatch": pixel_match})
# Source/runtime equality is recorded, not assumed from deterministic generation alone.
bomb_result = {"assets": assets, "bleedAlphaUnchanged": True, "productImagesWritten": 0}

# Check the existing cutout metadata without running its destructive top-level save.
cutout_path = COCOS / "art-source/ui/tournament-cutout.py"
compile(ast.parse(cutout_path.read_text()), str(cutout_path), "exec")
with Image.open(COCOS / "assets/game-assets/ui/lobby/entry-tournament-cutout-v4.png") as existing:
    assert existing.size == (512, 512) and existing.mode == "RGBA"
    assert existing.getchannel("A").getextrema() == (0, 255)
    cutout = {"size": existing.size, "mode": existing.mode, "alpha": [0, 255],
              "scriptExecuted": False}

# Audio CLI gets only new synthetic WAVs. No original voice or profile data is changed.
audio_script = COCOS / "scripts/prepare-steel-plate-voice.py"
audio_cases = []
env = {**os.environ, "PYTHONDONTWRITEBYTECODE": "1"}
def make_wav(path, channels=1, silent=False):
    rate = 24000
    samples = np.zeros(rate, dtype="<i2")
    if not silent:
        for i in range(6000, 18000):
            samples[i] = round(6000 * math.sin(2 * math.pi * 440 * i / rate))
    with wave.open(str(path), "wb") as stream:
        stream.setparams((channels, 2, rate, 0, "NONE", "not compressed"))
        stream.writeframes(np.repeat(samples, channels).astype("<i2").tobytes())
with tempfile.TemporaryDirectory(prefix="audit-audio36-") as td:
    scratch = Path(td)
    source, output = scratch / "input.wav", scratch / "out.wav"
    make_wav(source)
    source_hash = digest(source.read_bytes())
    result = subprocess.run([sys.executable, str(audio_script), str(source), str(output)],
                            capture_output=True, text=True, env=env, timeout=20)
    assert result.returncode == 0, result.stderr
    with wave.open(str(output), "rb") as stream:
        assert stream.getparams()[:3] == (1, 2, 24000)
        pcm = np.frombuffer(stream.readframes(stream.getnframes()), dtype="<i2")
        assert .25 < len(pcm)/24000 < 2.5
        assert pcm[0] == pcm[-1] == 0
        assert np.abs(pcm.astype(np.int32)).max() <= 32767 * 10**(-3/20) + 1
        audio_cases.append({"case": "syntheticSuccess", "duration": len(pcm)/24000,
                            "monoPcm16": True, "fadeEndpointsZero": True})
    output_hash = digest(output.read_bytes())
    retry = subprocess.run([sys.executable, str(audio_script), str(source), str(output)],
                           capture_output=True, text=True, env=env, timeout=20)
    assert retry.returncode != 0 and "Destination exists" in retry.stderr
    assert digest(output.read_bytes()) == output_hash and digest(source.read_bytes()) == source_hash
    audio_cases.append({"case": "existingDestinationRejected", "bothUnchanged": True})
    for name, kwargs, message in [
        ("silence", {"silent": True}, "Silent source"),
        ("stereo", {"channels": 2}, "AssertionError"),
    ]:
        src, dst = scratch / (name + ".wav"), scratch / (name + "-out.wav")
        make_wav(src, **kwargs)
        rejected = subprocess.run([sys.executable, str(audio_script), str(src), str(dst)],
                                  capture_output=True, text=True, env=env, timeout=20)
        assert rejected.returncode != 0 and message in rejected.stderr
        assert not dst.exists()
        audio_cases.append({"case": name + "Rejected", "noOutput": True})

print(json.dumps({"python": sys.version.split()[0], "pillow": Image.__version__,
                  "timer": timer_result, "bomb": bomb_result, "cutout": cutout,
                  "historicalDocument": doc, "audio": audio_cases,
                  "scope": "memory or owned temporary fixtures; no product write, TTS, network or build"}, indent=2))

File: docs/audit/2026-09-12/repro/asset-tools-36.mjs
// Audit-only: generators execute with captured writes; no network or product changes.
import assert from 'node:assert/strict'
import fs from 'node:fs'
import { createHash } from 'node:crypto'
import vm from 'node:vm'
import { deflateSync, inflateSync } from 'node:zlib'
import { fileURLToPath, pathToFileURL } from 'node:url'
import { dirname, resolve } from 'node:path'
const root=resolve(dirname(fileURLToPath(import.meta.url)),'../../../..')
const cocos=resolve(root,'work/guandan-cocos')
const sha=b=>createHash('sha256').update(b).digest('hex')
const text=p=>fs.readFileSync(p,'utf8')
const run=(path,context)=>vm.runInNewContext(text(path).replace(/^import .*$/gm,'').replaceAll('import.meta.url',JSON.stringify(pathToFileURL(path).href)),context,{filename:path,timeout:10000})
const writes=[],directories=[]
run(resolve(cocos,'scripts/generate-lobby-glint.mjs'),{
  Buffer, URL, fileURLToPath, deflateSync, console:{log(){}},
  mkdirSync:(p,o)=>directories.push([p,o]),
  writeFileSync:(p,data)=>writes.push({path:p,data:Buffer.from(data)}),
})
assert.equal(writes.length,1);assert.equal(directories.length,1)
assert.equal(writes[0].path,resolve(cocos,'assets/game-assets/effects/lobby-v1/button-glint.png'))
const png=writes[0].data
assert.deepEqual(png.subarray(0,8),Buffer.from([137,80,78,71,13,10,26,10]))
const chunks=[]
// Independent CRC implementation uses a table, not the source generator's bitwise loop per byte.
const table=Array.from({length:256},(_,v)=>{for(let k=0;k<8;k++)v=v&1?0xedb88320^(v>>>1):v>>>1;return v>>>0})
const crc=b=>{let v=0xffffffff;for(const byte of b)v=(v>>>8)^table[(v^byte)&255];return(v^0xffffffff)>>>0}
for(let i=8;i<png.length;){
  const length=png.readUInt32BE(i),kind=png.subarray(i+4,i+8).toString()
  const body=png.subarray(i+4,i+8+length)
  assert.equal(crc(body),png.readUInt32BE(i+8+length))
  chunks.push({kind,data:png.subarray(i+8,i+8+length)});i+=length+12
  assert.ok(i<=png.length)
}
assert.deepEqual(chunks.map(x=>x.kind),['IHDR','IDAT','IEND'])
const header=chunks[0].data
assert.equal(header.readUInt32BE(0),384);assert.equal(header.readUInt32BE(4),96)
assert.equal(header[8],8);assert.equal(header[9],6)
const raw=inflateSync(chunks[1].data), stride=384*4+1
assert.equal(raw.length,stride*96)
const frameMax=Array(16).fill(0)
for(let y=0;y<96;y++){
  assert.equal(raw[y*stride],0)
  for(let x=0;x<384;x++){
    const f=Math.floor(y/48)*8+Math.floor(x/48)
    frameMax[f]=Math.max(frameMax[f],raw[y*stride+1+x*4+3])
  }
}
assert.equal(frameMax[0],0);assert.equal(frameMax[15],0)
assert.ok(frameMax.slice(1,15).every(x=>x>0))
// sin() floating-point rounding can differ by one alpha unit in mirrored frames.
assert.ok(frameMax.every((value,index)=>Math.abs(value-frameMax[15-index])<=1))
const existing=fs.readFileSync(writes[0].path)
assert.deepEqual(png,existing)
const glint={size:[384,96],frames:16,bytes:png.length,sha256:sha(png),sourceMatchesExisting:true,chunkCrcChecks:3,frameAlphaMax:frameMax,productWrites:0}

const server=resolve(root,'work/guandan-windows-source/server/data/default-profiles')
const catalog=JSON.parse(text(resolve(server,'catalog.json')))
assert.equal(catalog.version,1)
assert.equal(catalog.profiles.length,24)
assert.equal(catalog.failures.length,26)
const ids=new Set(),names=new Set()
for(const p of catalog.profiles){
  assert.match(p.id,/^\d{3}$/);assert.match(p.file,/^\d{3}\.(jpg|png|gif)$/)
  assert.equal(p.file.slice(0,3),p.id)
  assert.equal(p.avatarUrl,'profile:'+p.id)
  assert.ok(p.displayName&&[...p.displayName].length<=24)
  assert.ok(!ids.has(p.id));ids.add(p.id);names.add(p.displayName)
  assert.deepEqual(fs.readFileSync(resolve(server,p.file)),fs.readFileSync(resolve(cocos,'assets/game-assets/ui/profiles',p.file)))
}
assert.deepEqual([...ids].map(Number).concat(catalog.failures).sort((a,b)=>a-b),Array.from({length:50},(_,i)=>i+1))
const copied=[],catalogWrites=[]
run(resolve(cocos,'scripts/sync-default-profiles.mjs'),{
  URL,console:{log(){}},mkdirSync(){},
  readFileSync:(p,encoding)=>{assert.equal(fileURLToPath(p),resolve(server,'catalog.json'));return fs.readFileSync(p,encoding)},
  copyFileSync:(s,d)=>copied.push([fileURLToPath(s),fileURLToPath(d)]),
  writeFileSync:(p,data)=>catalogWrites.push([fileURLToPath(p),data]),
})
assert.equal(copied.length,24);assert.equal(catalogWrites.length,1)
for(const [s,d] of copied){
  assert.equal(dirname(s),server)
  assert.equal(dirname(d),resolve(cocos,'assets/game-assets/ui/profiles'))
}
assert.equal(catalogWrites[0][0],resolve(cocos,'assets/scripts/services/DefaultProfileCatalog.ts'))
assert.equal(catalogWrites[0][1],text(catalogWrites[0][0]))
const profiles={successfulPairs:24,failedOrNotFetchedIndices:26,uniqueIds:ids.size,uniqueNames:names.size,completeIndexPartition:true,serverClientPixelFileHashesMatch:24,generatedTsMatchesActual:true,newQueries:0}
console.log(JSON.stringify({glint,profiles,scope:'read-only catalog; generator final writes captured in memory; no avatar fetching'},null,2))
